src/algo/maximum_flow/dinics.rs

```diff
@@ -0,0 +1,335 @@
+/// their distance to the source vertex, considering only edges with
+/// positive residual capacity.
+///
+/// The source vertex is labeled as 1, and vertices not reachable are
+/// labeled as 0.
+///
+/// Aggregates in `level_edges` the edges that connects each
+/// vertex to its neighbours in the next level.
+///
+/// Returns the computed level graph.
+pub fn build_level_graph<N>(
+    network: N,
+    source: N::NodeId,
+    destination: N::NodeId,
+    flows: &[N::EdgeWeight],
+    level_edges: &mut [Vec<N::EdgeRef>],
+) -> Vec<usize>
+where
+    N: NodeCount + IntoEdgesDirected + NodeIndexable + EdgeIndexable,
+    N::EdgeWeight: Sub<Output = N::EdgeWeight> + PositiveMeasure,
+{
+    let mut level_graph = vec![0; network.node_count()];
+    let mut bfs_queue = VecDeque::with_capacity(network.node_count());
+    bfs_queue.push_back(source);
+
+    level_graph[NodeIndexable::to_index(&network, source)] = 1;
+    while let Some(vertex) = bfs_queue.pop_front() {
+        let vertex_index = NodeIndexable::to_index(&network, vertex);
+        let out_edges = network.edges_directed(vertex, Direction::Outgoing);
+        let in_edges = network.edges_directed(vertex, Direction::Incoming);
+        level_edges[vertex_index].clear();
+        for edge in out_edges.chain(in_edges) {
+            let next_vertex = other_endpoint(&network, edge, vertex);
+            let edge_index = EdgeIndexable::to_index(&network, edge.id());
+            let residual_cap = residual_capacity(&network, edge, next_vertex, flows[edge_index]);
+            if residual_cap == N::EdgeWeight::zero() {
+                continue;
+            }
+            let next_vertex_index = NodeIndexable::to_index(&network, next_vertex);
+            if level_graph[next_vertex_index] == 0 {
+                level_graph[next_vertex_index] = level_graph[vertex_index] + 1;
+                level_edges[vertex_index].push(edge);
+                if next_vertex != destination {
+                    bfs_queue.push_back(next_vertex);
+                }
+            } else if level_graph[next_vertex_index] == level_graph[vertex_index] + 1 {
+                level_edges[vertex_index].push(edge);
+            }
+        }
+    }
+
+    level_graph
+}
+
+/// Find blocking flow for current level graph by repeatingly finding
+/// augmenting paths in it.
+///
+/// Attach computed flows to `flows` and returns the total flow increase from
+/// edges available in `level_edges` at this iteration.
+fn find_blocking_flow<N>(
+    network: N,
+    source: N::NodeId,
+    destination: N::NodeId,
+    flows: &mut [N::EdgeWeight],
+    level_edges: &mut [Vec<N::EdgeRef>],
+    visited: &mut N::Map,
+) -> N::EdgeWeight
+where
+    N: NodeCount + IntoEdges + NodeIndexable + EdgeIndexable + Visitable,
+    N::EdgeWeight: Sub<Output = N::EdgeWeight> + PositiveMeasure,
+{
+    let mut flow_increase = N::EdgeWeight::zero();
+    let mut edge_to = vec![None; network.node_count()];
```

Comment:
```suggestion
    let mut edge_to = vec![None; network.node_bound()];
```
See previous comments